contracts/serverless-v2/Jobs.sol

```diff
@@ -0,0 +1,456 @@
+
+    event JobTimeoutFailure(uint256 indexed jobId, bool callback_success);
+
+    error JobsRelayTimeOver();
+    error JobsInvalidSequenceId();
+    error JobsJobAlreadyRelayed();
+    error JobsUnsupportedChain();
+    error JobsSignatureTooOld();
+    error JobsExecutionTimeOver();
+    error JobsNotSelectedExecutor();
+    error JobsExecutorAlreadySubmittedOutput();
+    error JobsUnavailableResources();
+
+    //-------------------------------- internal functions start --------------------------------//
+
+    function _createJob(
+        bytes32 _codehash,
+        bytes memory _codeInputs,
+        uint256 _deadline, // in milliseconds
+        address _jobOwner
+    ) internal returns (uint256 jobId) {
+        address[] memory selectedNodes = EXECUTORS.selectExecutors(NO_OF_NODES_TO_SELECT);
+        // if no executors are selected, then return with error code 1
+        if (selectedNodes.length < NO_OF_NODES_TO_SELECT) {
+            revert JobsUnavailableResources();
+        }
+
+        // deposit escrow amount(USDC)
+        USDC_TOKEN.safeTransferFrom(
+            _jobOwner,
+            address(this),
+            _deadline * (EXECUTOR_FEE_PER_MS + STAKING_REWARD_PER_MS)
+        );
+
+        jobId = _create(_codehash, _codeInputs, _deadline, _jobOwner, selectedNodes);
+    }
+
+    function _create(
+        bytes32 _codehash,
+        bytes memory _codeInputs,
+        uint256 _deadline, // in milliseconds
+        address _jobOwner,
+        address[] memory _selectedNodes
+    ) internal returns (uint256 jobId) {
+        // create a struct
+        jobId = jobs.length;
+        jobs.push();
+        jobs[jobId].deadline = _deadline;
+        jobs[jobId].execStartTime = block.timestamp;
+        jobs[jobId].jobOwner = _jobOwner;
+        jobs[jobId].selectedExecutors = _selectedNodes;
+
+        emit JobCreated(jobId, _jobOwner, _codehash, _codeInputs, _deadline, _selectedNodes);
+    }
+
+    function _submitOutput(
+        bytes memory _signature,
+        uint256 _jobId,
+        bytes memory _output,
+        uint256 _totalTime,
+        uint8 _errorCode,
+        uint256 _signTimestamp
+    ) internal {
+        if (
+            (block.timestamp * 1000) >
+            (jobs[_jobId].execStartTime * 1000) + jobs[_jobId].deadline + (EXECUTION_BUFFER_TIME * 1000)
+        ) revert JobsExecutionTimeOver();
+
+        // signature check
+        address enclaveAddress = _verifyOutputSign(_signature, _jobId, _output, _totalTime, _errorCode, _signTimestamp);
+
+        if (!_isJobExecutor(_jobId, enclaveAddress)) revert JobsNotSelectedExecutor();
+        if (jobs[_jobId].hasExecutedJob[enclaveAddress]) revert JobsExecutorAlreadySubmittedOutput();
+
+        EXECUTORS.releaseExecutor(enclaveAddress);
+        jobs[_jobId].hasExecutedJob[enclaveAddress] = true;
+
+        uint8 outputCount = ++jobs[_jobId].outputCount;
+        if (outputCount == 1) jobs[_jobId].executionTime = _totalTime;
```

Comment:
probably cap it at `deadline` explicitly

Reply:
Okay, using the minimum of deadline and _totalTime. Not reverting if reported _totalTime is greater than deadline.